### DISCUSSION TITLE
Guide to creating a pipeline

### DISCUSSION THREAD

**Discussion by zabirauf** [open-webui/pipelines#189] [Show and tell]
I love OpenWebUI and glad we have the pipelines mechanism to extend it. The documentation on this is currently lacking so it does take some time to go over and understand various types of pipeline, when to create which, what are valves etc. I wrote a blog post on my understanding of various aspects for pipelines. I hope it helps someone who is writing a custom pipeline.

http://zohaib.me/extending-openwebui-using-pipelines/

**Comment by douo** [comment]
Thank you, that was very helpful!

**Comment by aouicher** [comment]
That's great !!!

**Reply by mylegitches** [reply]
Much obliged.

**Comment by Urvesh71** [comment]
Well, I saw the documentation you uploaded on http://zohaib.me/extending-openwebui-using-pipelines/. But this does not clarify the idea that How to create the Pipeline and enable it to use in Open Web UI. 

It would be better if you upload step by step points for the guideline.

**Comment by Urvesh71** [comment]
Kindly prepare the guideline how to enable the text_to_sql_pipeline.py in Open Web UI.

**Comment by sir3mat** [comment]
Hi, thanks for your effort!
I have some questions:
1) When I create a model by going to Workspaces -> Model -> Create a Model and selecting the model ID of a custom pipeline as the base model, the inlet and outlet functions don't seem to be triggered.
Any insights on why this might be happening? 
2) What is the best approach to delegate a Retrieval-Augmented Generation (RAG) process to a custom pipeline?
Context:
In this custom pipeline setup, the workflow is as follows:

    Documents are loaded and fed into OpenWebUI.
    OpenWebUI encodes the documents, splits them into chunks, and stores them.
    The /inlet endpoint on the pipeline is called.
    OpenWebUI then performs the RAG operation.
    The pipeline’s /pipe endpoint is triggered, displaying the response on OpenWebUI.
    Finally, the client calls the /outlet endpoint to retrieve the result.

Question on Implementation:
Why do the /inlet, /pipe, and /outlet endpoints each have different request bodies?

**Comment by Likeux** [comment]
A man takes his time to create something useful for anyone to use and people will use him like a chat bot to answer their questions instead of trouble shooting or trying to do something themselves. Some even DEMAND things to be added or changed.
